rust/template/distributed_datalog/src/txnmux.rs

```diff
@@ -0,0 +1,97 @@
+use std::any::Any;
+use std::fmt::Debug;
+
+use observe::CachingObserver;
+use observe::Observable;
+use observe::ObservableBox;
+use observe::ObserverBox;
+use observe::OptionalObserver;
+use observe::SharedObserver;
+
+/// A multiplexer for transactions. In a nutshell, this is an object
+/// that tracks a set of `Observable`s and implements the `Observable`
+/// interface (i.e., can be subscribed to) itself. It will ensure that
+/// the registered `Observable`'s transactions are serialized, meaning
+/// the concurrent transactions will be submitted one after the other
+/// without interleaving.
+#[derive(Debug, Default)]
+pub struct TxnMux<T, E>
+where
+    T: Debug + Send,
+    E: Debug + Send,
+{
+    /// The observables we track and our subscriptions to them.
+    subscriptions: Vec<(ObservableBox<T, E>, Box<dyn Any>)>,
+    /// A reference to the `Observer` subscribed to us, if any.
+    observer: SharedObserver<OptionalObserver<ObserverBox<T, E>>>,
+}
+
+impl<T, E> TxnMux<T, E>
+where
+    T: Debug + Send + 'static,
+    E: Debug + Send + 'static,
+{
+    /// Create a new `TxnMux`, without any observables.
+    pub fn new() -> Self {
+        Self {
+            subscriptions: Vec::new(),
+            observer: SharedObserver::new(OptionalObserver::default()),
+        }
+    }
+
+    /// Add an `Observable` to track to the multiplexer.
+    pub fn add_observable(
```

Comment:
Interesting, this is a new, non-standard way to connect `Observable` to an object that does not implement the `Observer` interface. A potential problem here is how the control plane will "know" how to instantiate such a graph. We can discuss in person.

Reply:
Right, I did not see a good way to fudge that with the observable/observer infrastructure. The control plane would need knowledge of this interface.

I think in the future we may want to have an "integrated" multiplexer. Basically, that could be a TCP endpoint that repeatedly instantiates an observable (`TcpReceiver`) for each accepted connection. So there would be less control plane involvement.

Reply:
Another approach is to have the multiplexer expose multiple observer instances. But this would still require a non-standard interface to allocate a new observer, so it's not a clear win to me.

Reply:
Let's discuss some more tomorrow.